lib/ui/painting.dart

```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
   /// Returns a copy of the path with all the segments of every
   /// subpath transformed by the given matrix.
   Path transform(Float64List matrix4) {
-    assert(matrix4 != null);
-    if (matrix4.length != 16)
-      throw new ArgumentError('"matrix4" must have 16 entries.');
+    assert(_matrix4IsValid(matrix4));
     return _transform(matrix4);
   }
   Path _transform(Float64List matrix4) native 'Path_transform';
+
+  /// Gets the bounding rectangle for this path.
+  Rect getBounds() {
+    final Float32List rect = _getBounds();
+    return new Rect.fromLTRB(rect[0], rect[1], rect[2], rect[3]);
+  }
+  Float32List _getBounds() native 'Path_getBounds';
+
+  /// Sets this path to the result of operation [PathOperation]
+  /// 
+  /// If only one path is specified, this path will be used as the first 
+  /// operand, otherwise path1 will be the first operand and path2 the
+  /// second.
+  /// 
+  /// The resulting path will be constructed from non-overlapping contours. The
+  /// curve order is reduced where possible so that cubics may be turned into
+  /// quadratics, and quadratics maybe turned into lines.
+  static Path combine(PathOperation operation, Path path1, Path path2) {
+    assert(path1 != null);
+    assert(path2 != null);
+
+    final Path path = new Path();
+    if (path._op(path1, path2, operation.index)) {
+      return path;
+    } 
+    throw new StateError('Path combine failed - likely due to invalid path');
+  }
+  bool _op(Path path1, Path path2, int operation) native 'Path_op';
+
+  /// Creates a [PathMetrics] object for this path.
+  PathMetrics computeMetrics({bool forceClosed = false}) {
+    return new PathMetrics._(this, forceClosed);
+  }
+}
+
+/// Convenience class to return the result of [PathMetric.getTangentForOffset].
+class Tangent {
+  /// Position of the tangent
+  final Offset position;
+  /// Angle of the tangent in radians
+  final double angle;
+
+  const Tangent(this.position, this.angle);
```

Comment:
Please put constructors at the top of the class, before the fields. That way it's easy to tell at a glance if there's an implicit constructor without reading the whole class.

Reply:
Also the constructor should have docs, something like:

```
/// Creates a [Tangent] with the given values.
///
/// The arguments must not be null.
```

...and then add initializer asserts that verify they're not null:

```dart
  /// Creates a [Tangent] with the given values.
  ///
  /// The arguments must not be null.
  const Tangent(this.position, this.angle) : assert(position != null), assert(angle != null);
 
```